protocol/communities/community_admin_event.go

```diff
@@ -0,0 +1,346 @@
+		Clock:       o.Clock(),
+		CommunityId: o.ID(),
+		Type:        protobuf.CommunityAdminEvent_COMMUNITY_EDIT,
+		CommunityConfig: &protobuf.CommunityConfig{
+			Identity:      o.Identity(),
+			Permissions:   o.Permissions(),
+			AdminSettings: o.AdminSettings(),
+			IntroMessage:  o.IntroMessage(),
+			OutroMessage:  o.OutroMessage(),
+			Tags:          o.TagsRaw(),
+		},
+	}
+}
+
+func (o *Community) ToCommunityBecomeMemberTokenPermissionChangeAdminEvent() *protobuf.CommunityAdminEvent {
+	tokenPermissions := make(map[string]*protobuf.CommunityTokenPermission)
+	for _, tp := range o.TokenPermissionsByType(protobuf.CommunityTokenPermission_BECOME_MEMBER) {
+		tokenPermissions[tp.Id] = tp
+	}
+	return &protobuf.CommunityAdminEvent{
+		Clock:            o.Clock(),
+		CommunityId:      o.ID(),
+		Type:             protobuf.CommunityAdminEvent_COMMUNITY_MEMBER_TOKEN_PERMISSION_CHANGE,
+		TokenPermissions: tokenPermissions,
+	}
+}
+
+func (o *Community) ToCommunityBecomeMemberTokenPermissionDeleteAdminEvent() *protobuf.CommunityAdminEvent {
+	tokenPermissions := make(map[string]*protobuf.CommunityTokenPermission)
+	for _, tp := range o.TokenPermissionsByType(protobuf.CommunityTokenPermission_BECOME_MEMBER) {
+		tokenPermissions[tp.Id] = tp
+	}
+	return &protobuf.CommunityAdminEvent{
+		Clock:            o.Clock(),
+		CommunityId:      o.ID(),
+		Type:             protobuf.CommunityAdminEvent_COMMUNITY_MEMBER_TOKEN_PERMISSION_DELETE,
+		TokenPermissions: tokenPermissions,
+	}
+}
+
+func (o *Community) ToCommunityRequestToJoinAcceptAdminEvent(changes *CommunityAdminEventChanges) *protobuf.CommunityAdminEvent {
+	return &protobuf.CommunityAdminEvent{
+		Clock:                  o.Clock(),
+		CommunityId:            o.ID(),
+		Type:                   protobuf.CommunityAdminEvent_COMMUNITY_REQUEST_TO_JOIN_ACCEPT,
+		MembersAdded:           changes.MembersAdded,
+		AcceptedRequestsToJoin: changes.AcceptedRequestsToJoin,
+	}
+}
+
+func (o *Community) ToCommunityRequestToJoinRejectAdminEvent(changes *CommunityAdminEventChanges) *protobuf.CommunityAdminEvent {
+	return &protobuf.CommunityAdminEvent{
+		Clock:                  o.Clock(),
+		CommunityId:            o.ID(),
+		Type:                   protobuf.CommunityAdminEvent_COMMUNITY_REQUEST_TO_JOIN_REJECT,
+		RejectedRequestsToJoin: changes.RejectedRequestsToJoin,
+	}
+}
+
+func (o *Community) PatchCommunityDescriptionByAdminEvent(adminEvent *protobuf.CommunityAdminEvent) (*protobuf.CommunityDescription, error) {
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	if adminEvent.Clock <= o.config.CommunityDescription.Clock {
+		return nil, errors.New("clock for admin event is outdated")
+	}
+
+	if adminEvent.Type == protobuf.CommunityAdminEvent_UNKNOWN {
+		return nil, errors.New("unknown admin event")
+	}
+
+	// create a deep copy of current community so we can patch CommunityDescription
+	copy := o.createDeepCopy()
+
+	copy.config.CommunityDescription.Clock = adminEvent.Clock
```

Comment:
I think this maybe needs to be adjusted, if we go for events, you don't want the last one wins, you should be able to process them regardless of the order.

just to make sure, if I read the code correctly, let's take this example:

1) `admin A` changes say the name of a channel at `clock` 3
2)   `admin B` changes the name of a different channel at `clock` 4.
3) Client `C` retrieves and process 2) and 1) in that order

Here, seems like 1) would be discarded, since the clock would be `4`, but the last received event has clock `3`

Is that correct?

I think the desired behavior is that both are processed

Reply:
Yes, but I was orienting on `func (o *Community) UpdateCommunityDescription`, where we do clock validation
Does it means, that if instead of 2 admins we as an example take 2 owners - if 2 owners send some update with the scenario above, we'll get the same result? So it is not an admin only issue, but a global one

Reply:
2 owners I think we are fine with `last one wins`, as we don't really expect to have multiple owners for community.

When it comes to admins, is a bit different, we do expect multiple admins to make changes to the community at different times

Reply:
Any suggestions? :)
Do not track the clock?

Reply:
but again, this can bring another issue, when admin action will increase the clock and the owner msg can be rejected later if he'll not receive an admin event before sending his msg...

Reply:
> Any suggestions? :) Do not track the clock? Just increase it to 1 from the current client CommunityDescription clock?

I think we need to separate clocks for owner and admins, for private group chats, you "build" the group chat from a stream of events, but you don't keep track of the clock in a different piece of state ( the clock of the events is the latest clock in the event stream).
So I would not mix the two, basically a community is:

CommunityDescription -> only modifiable by a owner, has its own clock, when receiving an admin update, this clock should not be modified
CommunityAdminEvents -> each one has its own clock, they should not modify directly the community description

AmendedCommunityDescription = CommunityDescription + Process Any Event with clock > CommunityDescription.Clock

The issue is the owner republishing the community description / making changes without having received all the admin events, that might result in some events being ignored/discarded, I don't think there's a way around it, but admin can always republish events if they see that's the case
That would be the solution I'd be leaning towards
This is sort of similar, you apply events on top of the latest community description.

> when admin action will increase the clock and the owner msg can be rejected later if he'll not receive an admin event before sending his msg...

I believe this should not happen, if we don't mix clocks